README.md
# Sapphire

Life-event detection algorithm and software.

![demoimage](./demoimage.png)

## Description

Dashboard application for viewing activity of drosophila and detecting life-events.

## Versions

You can find version information in [CHANGELOG.md](./CHANGELOG.md).

## Dependencies

The application requires Python 3.6.
It also needs Python modules as described bellow.

If you want to use only Sapphire (only viewing images, signals, etc):

| Module name | Version | Description |
| ---- | ----: | ---- |
| dash | 0.43.0 | You have not to install following modules because it is automatically installed with `dash`.<br> - `dash-core-components`<br> - `dash-html-components`<br> - `dash-renderer`<br> - `dash-table`<br> - `plotly`<br> If you use `conda`/`pip`, execute the install command specifying the version: `conda install -c conda-forge dash==0.43.0`/`pip install dash==0.43.0`. |
| dash-core-components | 0.48.0 |  |
| dash-html-components | 0.16.0 | |
| dash-renderer | 0.24.0 | |
| dash-table | 3.7.0 | |
| numpy | 1.16.5 | |
| pandas | 0.25.1 | |
| Pillow | 6.1.0 | |
| plotly | 3.2.0 | |
| scipy | 1.3.1 | |

If you want to carry out inference and/or further analysis with the scripts in this repository:

| Module name | Version | Description |
| ---- | ----: | ---- |
| changefinder | 0.3 | Please use `pip install` command if you are using Anaconda/Miniconda environment. This is not provided by the Anaconda repository at this point (2020/04/08). |
| cudatoolkit | 9.0 | Automatically installed by `tensorflow-gpu` installing. |
| cudnn | 7.6.0 | Automatically installed by `tensorflow-gpu` installing. |
| keras | 2.4.0 |  |
| tensorflow/tensorflow-gpu | 1.9.0 | You can acceralate inference by a neural network with GPU computing. In this case, please install `tensorflow-gpu`, not `tensorflow`. In installation of `tensorflow-gpu`, `cudatoolkit` and `cudnn` will automatically be installed. |
| tqdm | 4.32.1 |  |

This application run on any OS (Windows, macOS, Linux, etc.).
Running of Sapphire is verified in Windows 7, Ubuntu 16.04/18.04, and Apple OS X 10.11 El Capitan.
We have carried out operation verification of Sapphire in all the minor version releases.

## Usage

See the [manual](./manual_en.pdf) ([japanese ver.](./manual_ja.pdf)) in this repository.

## Install

Clone the source codes by executing below:  

``` shell
git clone git@github.com:kanglab/DiamondsOnDash.git
```

Or download the zip file by clicking the Clone or download button in this page.

Install Python language and the dependencies according to the Dependencies section (see the [manual](./manual_en.pdf) ([japanese ver.](./manual_ja.pdf))).
After installing all the dependencies, please execute `python` command in your shell terminal to start Sapphire:

``` shell
python sapphire.py
```

## Demo

This repository includes the directory `data_root_for_demo` for storing datasets.
Note that the directory tree under `data_root_for_demo` is regorously defined to load data appropriately.
Please put your data under the directories in `data_root_for_demo` without breaking the directory tree (renaming the directories is allowed).

And then, please open your web browser and access to `localhost:8050`.
Your browser will show Sapphire application.
If your installation is correct, you can select any parameters with the selector elements on Sapphire and explore flys' images, signals, and statistics like the [demo image](./demoimage.png).
If you can't select parameters (for example, a selector element has blank), Sapphire might raise some errors.
Please see your terminal to find erros and solve them carefully.

If your dataset has large number of frames (`original` directory has, for example, 10,000 images), Sapphire will take several seconds to show images, signals, and statistics on itself.

### Directory tree

The files/directories beggining with asterisk can be arbitrarily named, although the others should have the names as given below.

``` shell
*data_root_for_demo
├── *dataset1
│   ├── blacklist.csv
│   ├── config.json
│   ├── grouping.csv
│   ├── inference
│   │   ├── adult
│   │   │   └── *profile1
│   │   │       ├── *cf_r0.003_signals.npy
│   │   │       ├── probs
│   │   │       │   ├── 000.npz
│   │   │       │   ├── 001.npz
│   │   │       │   └── 002.npz
│   │   │       ├── probs.npz
│   │   │       └── *signals.npy
│   │   └── larva
│   │       └── *profile1
│   │           ├── probs
│   │           │   ├── 000.npz
│   │           │   ├── 001.npz
│   │           │   └── 002.npz
│   │           ├── probs.npz
│   │           └── *signals.npy
│   ├── mask.npy
│   ├── mask_params.json
│   ├── *network
│   │   ├── adult
│   │   │   ├── *profile1
│   │   │   └── *profile2
│   │   └── larva
│   │       ├── *profile1
│   │       └── *profile2
│   └── original
│       ├── 0001.jpg
│       ├── 0002.jpg
│       ├── 0003.jpg
│       ├── 0004.jpg
│       ├── 0005.jpg
│       ├── eclosion.csv
│       └── pupariation.csv
└── *dataset2
```

- `dataset1`, `dataset2`: Dataset directory.
- `blacklist.csv`: written in CSV (comma separated value) format.
- `config.json`: Configuration file for a dataset.
- `grouping.csv`: Defines groups of flies (CSV format).
- `inference`: Directory for storing results of inference by a trained neural network.
- `inference/adult` or `inference/larva`: Stores inference results for adult/larva flies.
- `inference/*/profile1`: The name of training profile indicating which trained network is used for the inference. The directory name is same as `network/*/profile1`.
- `inference/*/cf_r0.003_signals.npy`: ChangeFinder signal.
- `inference/*/probs`: Stores inference results of each fly in Numpy archive format. The number in file names indicates fly ID.
- `inference/*/probs.npz`: Numpy archive including inference results of all the flies.
- `inference/*/signals.npy`: Label diference signal.
- `mask.npy`: Definition of pixels of each fly in an original image. You can create this file with Sapphire's mask maker tab.
- `mask_params.json`: Parameters for creating the mask. You can create this file with Sapphire's mask maker tab.
- `network`: Stores neural networks trained for semantic segmentation.
- `network/adult` or `network/larva`: Stores networks for adult/larva flies.
- `network/*/profile1`: The name of training profile.
- `original`: Stores original images.

### Example network

We put pre-trained networks in the root of this repository. `pre_trained_network_for_adult.h5`/`pre_trained_network_for_larva.h5` is a network for detecting adult/larva animals. Pre-trained networks are expected to be located in a directory under `network` directory in `data_root_for_demo`. The `network` directory has `adult`/`larva` directory which should has directories for training profile. Please put `pre_trained_network_for_adult.h5` or `pre_trained_network_for_larva.h5` in the `network/adult/profile1` or `network/larva/profile2` directory, respectively.

## Licence

MIT license.  
If you use Sapphire for your work and find it useful, please cite our original paper ([Seong, et al., *eLife*, 2020](https://elifesciences.org/articles/58630), DOI: 10.7554/eLife.58630) in your paper.


CHANGELOG.md
# Change log

## [Ver. 0.2.0](https://github.com/kanglab/DiamondsOnDash/releases/tag/0.2.0) (2019-10-07)

**Implemented enhancements:**

- 閾値の算出方法を「全シグナルの平均値＋標準偏差x係数」から、シグナルごとの「最小値＋（最大値-最小値）/2」に変更しました。
- 羽化のみの判定に対応しました。
- イベント検出方法を2種類から選べるようにしました（極値を使った方法を追加しました）。
- 活動量シグナルのファイルを選択可能にしました。

**Bugfix:**

- シグナルグラフの選択データ点（赤点）が更新されないバグを修正しました。
- データセットを選択するごとにブラックリストを再読み込みするようにしました。

## [Ver. 0.2.1](https://github.com/kanglab/DiamondsOnDash/releases/tag/0.2.1) (2019-11-21)

**Bugfix:**

- データセットフォルダの名前に括弧（[]）が含まれる場合、globがうまく読み取らないので置換処理を加えました。

## [Ver. 0.3.0](https://github.com/kanglab/DiamondsOnDash/releases/tag/0.3.0) (2019-12-25)

**Implemented enhancements:**

- マスク作成用のタブを作成しました。マスクの作成と保存が可能です。
- 推論用スクリプト（inference.py）の追加
- ChangeFinder シグナル作成用スクリプト（make_CF_signals.py）の追加
- ファイルのリネーム（newviewer.py -> sapphire.py）
- 不要ファイルの削除（mask.py、viewer.py）

## [Ver. 0.3.1](https://github.com/kanglab/DiamondsOnDash/releases/tag/0.3.1) (2020-04-08)

**Modification:**

- Update README.
- Add manual_ja.pdf and manual_en.pdf.
- Add directory tree for demo.

## [Ver. 0.3.2](https://github.com/kanglab/DiamondsOnDash/releases/tag/0.3.2) (2021-06-21)

**Modification:**

- Add example pre-trained networks and explanation about it on README.
- Add the link to the original paper in the license statement.

**Implemented enhancements:**

- Do upsampling before inferencing if the size of an input image is less than 56x56.


inference.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2019-12-02
#
# Copyright (C) 2019 Taishi Matsumura
#
import os
import argparse

parser = argparse.ArgumentParser(description='Inference.')

parser.add_argument('inference_dataset_path',
        type=str, help='Put a path to a target dataset.')

parser.add_argument('trained_network_path',
        type=str, help='Put a path to a trained network.')

parser.add_argument('-g', '--gpu',
        type=int, help='Put an ID of GPU.')

args = parser.parse_args()

assert os.path.exists(args.inference_dataset_path),  \
        'The given path does not exists.'

assert os.path.exists(args.trained_network_path),  \
        'The given path does not exists.'

inference_dataset_path = args.inference_dataset_path
trained_network_path = args.trained_network_path
gpu_id = args.gpu

# 訓練済みネットワークがあるディレクトリの名前（target_dir）
target_dir = os.path.basename(os.path.dirname(trained_network_path))

# target_dir があるディレクトリの名前（morpho）
morpho = os.path.basename(os.path.dirname(os.path.dirname(trained_network_path)))

# 利用する GPU の ID を設定する
os.environ['CUDA_VISIBLE_DEVICES'] = str(gpu_id)

import PIL
import json
import glob
import numpy as np
from PIL import Image
from tqdm import tqdm

import keras
from keras import backend as K
from keras.utils import to_categorical


def TP(y_true, y_pred):
    y_true = K.argmax(y_true, axis=-1)
    y_pred = K.argmax(y_pred, axis=-1)
    return K.sum(K.cast(K.equal(y_true * y_pred, 1), K.floatx()))


def TN(y_true, y_pred):
    y_true = K.argmax(y_true, axis=-1)
    y_pred = K.argmax(y_pred, axis=-1)
    return K.sum(K.cast(K.equal(y_true + y_pred, 0), K.floatx()))


def FP(y_true, y_pred):
    y_true = K.argmax(y_true, axis=-1)
    y_pred = K.argmax(y_pred, axis=-1)
    return K.sum(K.cast(K.less(y_true, y_pred), K.floatx()))


def FN(y_true, y_pred):
    y_true = K.argmax(y_true, axis=-1)
    y_pred = K.argmax(y_pred, axis=-1)
    return K.sum(K.cast(K.greater(y_true, y_pred), K.floatx()))


def IoU(y_true, y_pred):
    return TP(y_true, y_pred)  \
            / (FN(y_true, y_pred) + TP(y_true, y_pred) + FP(y_true, y_pred))


def tf_normalize(images):
    images = images / 127.5
    return images - 1.0


def zeropadding(img, shape=(100, 100)):
    '''
    Input
    -----
    img : ndarray (height, width, n_channels)

    Output
    ------
    result : ndarray (height, width, n_channels)
    '''
    assert len(img.shape) == 2 or len(img.shape) == 3, 'Image must be 2D or 3D.'
    #assert img.shape[0] < shape[0], 'Image is too small.'
    #assert img.shape[1] < shape[1], 'Image is too small.'

    r0, c0 = int((shape[0] - img.shape[0]) / 2), int((shape[1] - img.shape[1]) / 2)
    if len(img.shape) == 3:
        result = np.zeros((shape[0], shape[1], img.shape[2]))
        result[r0:img.shape[0]+r0, c0:img.shape[1]+c0, :] = img
    elif len(img.shape) == 2:
        result = np.zeros((shape[0], shape[1]))
        result[r0:img.shape[0]+r0, c0:img.shape[1]+c0] = img
    return result


def split(image_path, n_wells, mask):
    org_image = np.array(PIL.Image.open(image_path).convert('L'), dtype=np.uint8)
    
    well_images = []
    for well_idx in range(n_wells):
        # Cut out a well image from the original image
        r, c = np.where(mask == well_idx)
        well_image = org_image[r.min():r.max(), c.min():c.max()]
		# Up sampling
        img = Image.fromarray(np.uint8(well_image))
        well_image = np.asarray(img.resize((56,56)))
        height, width = well_image.shape
        well_images.append(well_image)
        
    return np.array(well_images)


def get_well_imgs(img, mask, n_wells):
    well_imgs = []
    for i in range(n_wells):
        rows, clms = np.where(mask == i)
        well_imgs.append(img[min(rows):max(rows), min(clms):max(clms)])
    well_imgs = np.array(well_imgs)

    return well_imgs.reshape(
            well_imgs.shape[0], well_imgs.shape[1], well_imgs.shape[2], 1)


def inference(orgimg_path):
    # ウェル画像の切り出し
    well_images = split(orgimg_path, n_wells, mask)
    
    # (n_wells, height, width) -> (n_wells, height, width, 1)
    well_images = np.expand_dims(well_images, axis=-1)
    
    # Zeropadding
    well_images = np.array(
            [zeropadding(well_image, (56, 56)) for well_image in well_images])

    # データの正規化
    well_images = tf_normalize(well_images)

    # Inference
    return (100 * model.predict(well_images)).astype(np.uint8)


# ============
#  推論
# ============
# マスクのロード
mask = np.load(os.path.join(inference_dataset_path, 'mask.npy'))

# ウェル数のロード
with open(os.path.join(inference_dataset_path, 'mask_params.json')) as f:
    params = json.load(f)
n_wells = params['n-rows'] * params['n-clms'] * params['n-plates']

# 訓練済みネットワークのロード
model = keras.models.load_model(trained_network_path,
        custom_objects={'IoU': IoU, 'TP': TP, 'FP': FP, 'FN': FN, 'TN': TN})

# オリジナル画像のファイルパス
orgimg_paths = sorted(glob.glob(os.path.join(
        glob.escape(inference_dataset_path), 'original', '*.jpg')))

probs = []
# for each frame
for orgimg_path in tqdm(orgimg_paths):
    probs.append(inference(orgimg_path))
    

# ==============
#  データの作成
# ==============
# Tranform float probs into percentage
probs = np.array(probs).reshape(len(orgimg_paths), n_wells, 56, 56, 2)

# only probs which the animal is in the pixel
probs = probs[:, :, :, :, 1]

# Wells first (wells, org_imgs, height, width)
probs = probs.transpose(1, 0, 2, 3)

# シグナルの作成
signals = (np.diff((probs >= 50).astype(np.int8), axis=1)**2).sum(-1).sum(-1).T


# ==============
#  保存
# ==============
out_dir = os.path.join(inference_dataset_path, 'inference', morpho, target_dir)
os.makedirs(os.path.join(out_dir, 'probs'), exist_ok=True)

# probs の保存
for well_idx in range(n_wells):
    np.savez_compressed(os.path.join(
        out_dir, 'probs', '{:03d}.npz'.format(well_idx)), probs[well_idx])

np.savez_compressed(os.path.join(out_dir, 'probs.npz'), probs)

# シグナルの保存
np.save(os.path.join(out_dir, 'signals.npy'), signals)


viewer2.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2018-09-08
#
# Copyright (C) 2018 Taishi Matsumura
#
import io
import os
import glob
import dash
import base64
import PIL.Image
import dash_auth
import numpy as np
import scipy.ndimage
import flask_caching
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output, State


DATA_ROOT = 'G:/Research/Drosophila/CUI/data/'
THETA = 50


# -------
#  Main
# -------
app = dash.Dash()
app.css.append_css(
        {'external_url': 'https://codepen.io/chriddyp/pen/bWLwgP.css'})
cache = flask_caching.Cache()
cache.init_app(
        app.server, config={'CACHE_TYPE': 'filesystem', 'CACHE_DIR': 'cache/'})

app.layout = html.Div([
    html.Header([html.H1('Viewer')]),
    html.Div([
        html.Div([
            html.Div(['Image at "t"'], style={'display': 'table'}),
            html.Img(
                id='t-image',
                style={
                    'background': '#555555',
                    'height': '120px',
                    'width': '120px',
                    'padding': '5px',
                    'display': 'block',
                },
            ),
            html.Img(
                id='t-label',
                style={
                    'background': '#555555',
                    'height': '120px',
                    'width': '120px',
                    'padding': '5px',
                    'display': 'block',
                },
            ),
            ],
            style={
                'display': 'inline-block',
                'margin': '10px 10px',
            },
        ),
        html.Div([
            html.Div(['"t+1"'], style={'display': 'table'}),
            html.Img(
                id='t+1-image',
                style={
                    'background': '#555555',
                    'height': '120px',
                    'width': '120px',
                    'padding': '5px',
                    'display': 'block',
                },
            ),
            html.Img(
                id='t+1-label',
                style={
                    'background': '#555555',
                    'height': '120px',
                    'width': '120px',
                    'padding': '5px',
                    'display': 'block',
                },
            ),
            ],
            style={
                'display': 'inline-block',
                'margin': '10px',
            },
        ),
        html.Div([
            'Imaging environment :',
            html.Br(),
            html.Div([
                dcc.Dropdown(
                    id='env-dropdown',
                    placeholder='Select imaging env...',
                    clearable=False,
                ),
                ],
                style={
                    'display': 'inline-block',
                    'width': '300px',
                    'vertical-align': 'middle',
                },
            ),
            html.Br(),
            'CSV file :',
            html.Br(),
            html.Div([
                dcc.Dropdown(
                    id='csv-dropdown',
                    placeholder='Select CSV file...',
                    clearable=False,
                ),
                ],
                style={
                    'display': 'inline-block',
                    'width': '300px',
                    'vertical-align': 'middle',
                },
            ),
            html.Br(),
            'Morpho :',
            html.Br(),
            html.Div([
                dcc.Dropdown(
                    id='morpho-dropdown',
                    placeholder='Select morpho...',
                    clearable=False,
                ),
                ],
                style={
                    'display': 'inline-block',
                    'width': '300px',
                    'vertical-align': 'middle',
                },
            ),
            html.Br(),
            'Target to detect :',
            html.Br(),
            html.Div([
                dcc.Dropdown(
                    id='target-dropdown',
                    options=[
                        {'label': 'rise', 'value': 'rise'},
                        {'label': 'fall', 'value': 'fall'},
                    ],
                    value='rise',
                    placeholder='Detect...',
                    clearable=False,
                ),
                ],
                style={
                    'width': '100px',
                },
            ),
            'Well index :',
            html.Br(),
            html.Div([
                dcc.Input(
                    id='well-selector',
                    type='number',
                    value=0,
                    min=0,
                    size=5,
                ),
                ],
                style={'display': 'inline-block'},
            ),
            html.Div([
                dcc.Slider(
                    id='well-slider',
                    value=0,
                    min=0,
                    step=1,
                ),
                ],
                style={
                    'display': 'inline-block',
                    'width': '300px',
                    'margin-left': '20px',
                },
            ),
            html.Br(),
            'Time :',
            html.Br(),
            dcc.Input(
                id='time-selector',
                type='number',
                value=0,
                min=0,
                size=5,
            ),
            ],
            style={
                'display': 'inline-block',
                'margin': '10px 10px',
            },
        ),
        html.Div([
            'Data root :',
            html.Div(DATA_ROOT, id='data-root'),
            html.Br(),
            'Imaging environment :',
            html.Div(id='current-env'),
            html.Br(),
            'File name :',
            html.Div(id='current-csv'),
            html.Br(),
            'Current morpho :',
            html.Div(id='current-morpho'),
            ],
            style={
                'display': 'inline-block',
                'vertical-align': 'top',
                'margin': '10px 10px',
            },
        ),
        ],
    ),
    html.Div([
        dcc.Graph(
            id='signal-graph',
            style={
                'display': 'inline-block',
                'height': '500px',
                'width': '60%',
            },
        ),
        html.Div([
            dcc.Slider(
                id='threshold-slider',
                value=20,
                min=0,
                max=300,
                step=1,
                updatemode='mouseup',
                vertical=True,
            )],
            style={
                'display': 'inline-block',
                'height': '300px',
                'width': '5%',
                'padding-bottom': '100px',
            },
        ),
        dcc.Graph(
            id='summary-graph',
            style={
                'display': 'inline-block',
                'height': '500px',
                'width': '35%',
            },
        ),
    ]),
    ],
    style={
        'width': '1200px',
    },
)


@app.callback(
        Output('env-dropdown', 'options'),
        [Input('data-root', 'children')])
def callback(data_root):
    imaging_envs = [os.path.basename(i)
            for i in sorted(glob.glob(os.path.join(data_root, '*')))]

    return [{'label': i, 'value': i} for i in imaging_envs]


'''
@app.callback(
        Output('env-dropdown', 'value'),
        [Input('env-dropdown', 'options')])
def callback(envs):
    return envs[-1]['value']
'''


@app.callback(
        Output('csv-dropdown', 'options'),
        [Input('env-dropdown', 'value')],
        [State('data-root', 'children')])
def callback(env, data_root):
    if env is None:
        return []

    csvs = [os.path.basename(i)
            for i in sorted(glob.glob(os.path.join(
                data_root, env, 'original', '*.csv')))]

    return [{'label': i, 'value': i} for i in csvs]


@app.callback(
        Output('morpho-dropdown', 'options'),
        [Input('env-dropdown', 'value')],
        [State('data-root', 'children')])
def callback(env, data_root):
    if env is None:
        return []

    npys = [os.path.basename(i)
            for i in sorted(glob.glob(os.path.join(
                data_root, env, 'inference', '*'))) if os.path.isdir(i)]

    return [{'label': i, 'value': i} for i in npys]


@cache.memoize()
def store_signals(data_root, env, morpho):
    if env is None or morpho is None:
        return

    signals = np.load(os.path.join(
        data_root, env, 'inference', morpho, 'signals.npy'))

    return signals


@cache.memoize()
def store_manual_evals(data_root, env, csv):
    if env is None or csv is None:
        return

    manual_evals = np.loadtxt(
            os.path.join(data_root, env, 'original', csv),
            dtype=np.uint16,
            delimiter=',').flatten()

    return manual_evals


@cache.memoize()
def store_mask(data_root, env):
    if env is None:
        return

    return np.load(os.path.join(data_root, env, 'mask.npy'))



@app.callback(
        Output('current-morpho', 'children'),
        [Input('csv-dropdown', 'value'),
         Input('morpho-dropdown', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value')])
def callback(csv, morpho, data_root, env):
    if env is None or csv is None or morpho is None:
        return

    store_signals(data_root, env, morpho)
    store_manual_evals(data_root, env, csv)
    store_mask(data_root, env)
    return morpho


@app.callback(
        Output('current-env', 'children'),
        [Input('current-morpho', 'children')],
        [State('env-dropdown', 'value')])
def callback(_, env):
    return env


@app.callback(
        Output('current-csv', 'children'),
        [Input('current-morpho', 'children')],
        [State('csv-dropdown', 'value')])
def callback(_, csv):
    return csv


@app.callback(
        Output('time-selector', 'max'),
        [Input('current-morpho', 'children')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(_, data_root, env, morpho):
    if env is None or morpho is None:
        return

    signals = store_signals(data_root, env, morpho)
    return signals.shape[1] - 1


@app.callback(
        Output('threshold-slider', 'max'),
        [Input('current-morpho', 'children')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(_, data_root, env, morpho):
    if env is None or morpho is None:
        return

    signals = store_signals(data_root, env, morpho)
    return signals.max()


@app.callback(
        Output('well-slider', 'max'),
        [Input('current-morpho', 'children')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(_, data_root, env, morpho):
    if env is None or morpho is None:
        return

    signals = store_signals(data_root, env, morpho)
    return len(signals) - 1


@app.callback(
        Output('well-selector', 'max'),
        [Input('current-morpho', 'children')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(_, data_root, env, morpho):
    if env is None or morpho is None:
        return

    signals = store_signals(data_root, env, morpho)
    return len(signals) - 1


@app.callback(
        Output('well-slider', 'value'),
        [Input('summary-graph', 'clickData'),
         Input('current-morpho', 'children')])
def callback(click_data, _):
    if click_data is None:
        return 20

    return click_data['points'][0]['pointNumber']


@app.callback(
        Output('well-selector', 'value'),
        [Input('well-slider', 'value')])
def callback(well_idx):
    return well_idx


@app.callback(
        Output('time-selector', 'value'),
        [Input('signal-graph', 'clickData')])
def callback(click_data):
    if click_data is None:
        return 0
    else:
        return click_data['points'][0]['x']


@app.callback(
        Output('signal-graph', 'figure'),
        [Input('well-selector', 'value'),
         Input('threshold-slider', 'value'),
         Input('target-dropdown', 'value'),
         Input('time-selector', 'value')],
        [State('signal-graph', 'figure'),
         State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('csv-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(well_idx, threshold, rise_or_fall, time,
        figure, data_root, env, csv, morpho):
    if env is None or csv is None or morpho is None:
        return {'data': []}

    if len(figure['data']) == 0:
        x, y = 0, 0
    else:
        x, y = time, figure['data'][2]['y'][time]

    signals = store_signals(data_root, env, morpho)
    manual_evals = store_manual_evals(data_root, env, csv)

    if rise_or_fall == 'rise':
        auto_evals = (signals > threshold).argmax(axis=1)
    elif rise_or_fall == 'fall':
        auto_evals = signals.shape[1] - (np.fliplr(signals) > threshold).argmax(axis=1)
    return {
            'data': [
                {
                    # Manual evaluation time (vertical line)
                    'x': [manual_evals[well_idx]] * int(signals.max()),
                    'y': list(range(256)),
                    'mode': 'lines',
                    'name': 'Manual',
                    'line': {'width': 5, 'color': '#2ca02c'},
                },
                {
                    # Auto evaluation time (vertical line)
                    'x': [auto_evals[well_idx]] * int(signals.max()),
                    'y': list(range(256)),
                    'mode': 'lines',
                    'name': 'Auto',
                    'line': {'width': 5, 'color': 'd62728'},
                },
                {
                    # Signal
                    'x': list(range(len(signals[0, :]))),
                    'y': list(signals[well_idx]),
                    'mode': 'markers+lines',
                    'marker': {'size': 5, 'color': '#1f77b4'},
                    'name': 'Signal',
                },
                {
                    # Threshold (hrizontal line)
                    'x': list(range(len(signals[0, :]))),
                    'y': [threshold]*len(signals[0, :]),
                    'mode': 'lines',
                    'name': 'Threshold',
                    'line': {'width': 5, 'color': '#ff7f0e'},
                },
                {
                    # Selected data point
                    'x': [x],
                    'y': [y],
                    'mode': 'markers',
                    'marker': {'size': 10},
                    'name': '',
                },
            ],
            'layout': {
                'title': '',
                'font': {'size': 25},
                'xaxis': {
                    'title': 'Time step',
                    'tickfont': {'size': 25},
                },
                'yaxis': {
                    'title': 'Signal intensity',
                    'tickfont': {'size': 25},
                },
                'showlegend': False,
                'hovermode': 'closest',
            },
        }


@app.callback(
        Output('summary-graph', 'figure'),
        [Input('threshold-slider', 'value'),
         Input('well-selector', 'value'),
         Input('target-dropdown', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('csv-dropdown', 'value'),
         State('morpho-dropdown', 'value')])
def callback(threshold, well_idx, rise_or_fall, data_root, env, csv, morpho):
    if env is None or csv is None or morpho is None:
        return {'data': []}

    signals = store_signals(data_root, env, morpho)
    manual_evals = store_manual_evals(data_root, env, csv)

    if rise_or_fall == 'rise':
        auto_evals = (signals > threshold).argmax(axis=1)
    elif rise_or_fall == 'fall':
        # Scan the signal from the right hand side.
        auto_evals = signals.shape[1] - (np.fliplr(signals) > threshold).argmax(axis=1)
        # If the signal was not more than the threshold.
        auto_evals[auto_evals == signals.shape[1]] = 0

    return {
            'data': [
                {
                    'x': list(auto_evals),
                    'y': list(manual_evals),
                    'mode': 'markers',
                    'marker': {'size': 5},
                    'name': 'Well',
                },
                {
                    'x': [0, len(signals[0, :])],
                    'y': [0, len(signals[0, :])],
                    'mode': 'lines',
                    'name': 'Auto = Manual',
                },
                {
                    'x': [auto_evals[well_idx]],
                    'y': [manual_evals[well_idx]],
                    'mode': 'markers',
                    'marker': {'size': 10},
                    'name': 'Selected well',
                },
            ],
            'layout': {
                'title': 'Auto vs Manual',
                'xaxis': {
                    'title': 'Auto',
                },
                'yaxis': {
                    'title': 'Manual',
                },
                'hovermode': 'closest',
            },
        }


@app.callback(
        Output('t-image', 'src'),
        [Input('time-selector', 'value'),
         Input('well-selector', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value')])
def callback(time, well_idx, data_root, env):
    if env is None:
        return ''

    mask = store_mask(data_root, env)
    orgimg_paths = sorted(glob.glob(
            os.path.join(data_root, env, 'original', '*.jpg')))
    org_img = np.array(
            PIL.Image.open(orgimg_paths[time]).convert('L'), dtype=np.uint8)
    r, c = np.where(mask == well_idx)
    org_img = org_img[r.min():r.max(), c.min():c.max()]
    org_img = PIL.Image.fromarray(org_img)
    buf = io.BytesIO()
    org_img.save(buf, format='JPEG')
    return 'data:image/jpeg;base64,{}'.format(
            base64.b64encode(buf.getvalue()).decode('utf-8'))


@app.callback(
        Output('t+1-image', 'src'),
        [Input('time-selector', 'value'),
         Input('well-selector', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value')])
def callback(time, well_idx, data_root, env):
    if env is None:
        return ''

    mask = store_mask(data_root, env)
    orgimg_paths = sorted(glob.glob(
            os.path.join(data_root, env, 'original', '*.jpg')))
    org_img = np.array(
            PIL.Image.open(orgimg_paths[time+1]).convert('L'), dtype=np.uint8)
    r, c = np.where(mask == well_idx)
    org_img = org_img[r.min():r.max(), c.min():c.max()]
    org_img = PIL.Image.fromarray(org_img)
    buf = io.BytesIO()
    org_img.save(buf, format='JPEG')
    return 'data:image/jpeg;base64,{}'.format(
            base64.b64encode(buf.getvalue()).decode('utf-8'))


@app.callback(
        Output('t-label', 'src'),
        [Input('time-selector', 'value'),
         Input('well-selector', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('current-morpho', 'children')])
def callback(time, well_idx, data_root, env, morpho):
    if env is None or morpho is None:
        return ''

    label_images = sorted(glob.glob(
            os.path.join(data_root, env, 'inference',
                morpho, '{:03d}'.format(well_idx), '*.png')))
    buf = io.BytesIO()
    PIL.Image.open(label_images[time]).save(buf, format='PNG')
    return 'data:image/png;base64,{}'.format(
            base64.b64encode(buf.getvalue()).decode('utf-8'))


@app.callback(
        Output('t+1-label', 'src'),
        [Input('time-selector', 'value'),
         Input('well-selector', 'value')],
        [State('data-root', 'children'),
         State('env-dropdown', 'value'),
         State('current-morpho', 'children')])
def callback(time, well_idx, data_root, env, morpho):
    if env is None or morpho is None:
        return ''

    label_images = sorted(glob.glob(
            os.path.join(data_root, env, 'inference',
                morpho, '{:03d}'.format(well_idx), '*.png')))
    buf = io.BytesIO()
    PIL.Image.open(label_images[time+1]).save(buf, format='PNG')
    return 'data:image/png;base64,{}'.format(
            base64.b64encode(buf.getvalue()).decode('utf-8'))


if __name__ == '__main__':
    app.run_server(debug=True)
